File: Быстрая сортировка/quick sort.cpp
#define INCREASING <= // сортировка по возрастанию
#define DESCREASING >= // сортировка по убыванию

void quick_sort(std::vector<int>& massiv, int left, int right) {
    if (left > right) return;
    int wall = left; // создаём левую стену, где последний элемент стены - самый первый элемент массива.
    int pivot_index = right; // создаём индекс элемент, относительно которого мы будем сортировать числа
    int pivot_value = massiv[pivot_index]; // создаём элемент, относительно которого мы будем сортировать.
    int current_pos = left; // указываем текущую позицию указателя.

    for (; current_pos < right; current_pos++) {
        if (massiv[current_pos] INCREASING pivot_value) { // если значение указателя меньше значения pivot'a
            std::swap(massiv[current_pos], massiv[wall]); // то мы меняем местами последний элемент стены (wall) и элемент указателя.
            wall++; // достраиваем стену
        }
    }
    // все элементы, которые меньше или равны опорному элементу, расположены в стене, а которые больше - за пределами стены
    std::swap(massiv[wall], massiv[pivot_index]); // т.к элементы меньше опорного расположены в стене, то мы меняем местами элемент конца стены и опорный элемент
    quick_sort(massiv, left, wall - 1); // в параметры не включаем конец стены, потому что он уже расположен в правильной позиции
    quick_sort(massiv, wall + 1, right); // (ВСЕ элементы слева <= опорному, ВСЕ элементы справа > опорного)

File: Быстрая сортировка/quick-sort.js
'use strict'

/*
 It is my first time, when i start use JavScript and probably this program can be more simplify.
*/


let massiv = [34,31,6,1,67,14,623];


function swap(array, index1, index2) {
    [array[index1], array[index2]] = [array[index2], array[index1]];
}

function fast_sort(array, start=0, end=array.length-1) {
    if (start < end) {
        let pivot_value = massiv[end];
        let wall = start;
        for (let index = start; index < end; index++) {
            if (array[index] < pivot_value) {
                swap(array, index, wall);
                wall++;
            }
        }
        swap(array, wall, end);
        fast_sort(array, 0, wall-1);
        fast_sort(array, wall+1, end);
    }
}

fast_sort(massiv);
console.log(massiv);